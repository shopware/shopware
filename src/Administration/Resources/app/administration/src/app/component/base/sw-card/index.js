import template from './sw-card.html.twig';
import './sw-card.scss';

const { Component } = Shopware;

/**
 * @public
 * @description A card is a flexible and extensible content container.
 * @status ready
 * @example-type dynamic
 * @component-example
 * <sw-card title="Test title">
 *     Your content
 * </sw-card>
 */
Component.register('sw-card', {
    template,

    props: {
        title: {
            type: String,
            required: false
        },
        hero: {
            type: Boolean,
            required: false,
            default: false
        },
        isLoading: {
            type: Boolean,
            required: false,
            default: false
        },
        large: {
            type: Boolean,
            required: false,
            default: false
        }
    },

    computed: {
        cardClasses() {
            return {
                'sw-card--tabs': !!this.$slots.tabs || !!this.$scopedSlots.tabs,
                'sw-card--grid': !!this.$slots.grid || !!this.$scopedSlots.grid,
                'sw-card--hero': !!this.hero,
                'sw-card--large': this.large,
                'has--header': !!this.$slots.toolbar || !!this.$scopedSlots.toolbar
            };
        }
    }
});
