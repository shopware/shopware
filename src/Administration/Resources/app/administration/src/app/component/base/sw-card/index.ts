import template from './sw-card.html.twig';

const { Component } = Shopware;

/**
 * @package admin
 *
 * @private
 * @status ready
 * @description Wrapper component for sw-card and mt-card. Autoswitches between the two components.
 */
Component.register('sw-card', {
    template,

    compatConfig: {
        ...Shopware.compatConfig,
        INSTANCE_ATTRS_CLASS_STYLE: false,
    },

    computed: {
        useMeteorComponent() {
            // Use new meteor component in major
            if (Shopware.Feature.isActive('v6.7.0.0')) {
                return true;
            }

            // Throw warning when deprecated component is used
            Shopware.Utils.debug.warn(
                'sw-card',
                // eslint-disable-next-line max-len
                'The old usage of "sw-card" is deprecated and will be removed in v6.7.0.0. Please use "mt-card" instead.',
            );

            return false;
        },

        listeners() {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access
            if (this.isCompatEnabled('INSTANCE_LISTENERS')) {
                return this.$listeners;
            }

            return {};
        },
    },

    methods: {
        getSlots() {
            // eslint-disable-next-line @typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access
            if (this.isCompatEnabled('INSTANCE_SCOPED_SLOTS')) {
                return {
                    ...this.$slots,
                    ...this.$scopedSlots,
                };
            }

            return this.$slots;
        },
    },
});
