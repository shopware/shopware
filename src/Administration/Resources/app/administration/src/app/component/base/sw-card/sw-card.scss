@import "~scss/variables";

.sw-card {
    max-width: 800px;
    margin: 0 auto 40px auto;
    position: relative;
    color: $color-darkgray-200;

    &.sw-card--tabs {
        .sw-card__content {
            border-radius: 0 0 $border-radius-default $border-radius-default;
            border-top: none;
        }

        & > .sw-tabs {
            border: 1px solid $color-gray-300;
            border-radius: $border-radius-default $border-radius-default 0 0;
            border-bottom: none;
            margin-bottom: 0;

            .sw-tabs__content {
                background-color: $color-gray-50;
            }

            .sw-tabs__custom-Content {
                padding: 0;
            }

            .sw-tabs__arrow {
                top: (56px / 2);
                padding: 0 12px 0 12px;
            }

            &.sw-tabs--scrollable {
                padding: 0 (20px + 12px) 0 (20px + 12px);
            }

            .sw-tabs-item {
                padding-top: 19px;
                padding-bottom: 15px;
                font-weight: normal;

                &.sw-tabs-item--active {
                    font-weight: 600;
                }
            }
        }
    }

    &.sw-card--grid {
        .sw-card__content {
            display: grid;
            padding: 0;

            .sw-grid {
                border-top: none;
            }
        }
    }

    &.sw-card--hero {
        .sw-card__content {
            background: none;
            border: none;
            text-align: center;

            h3 {
                font-size: 30px;
            }
        }
    }

    &.sw-card--large {
        max-width: 1330px;

        .sw-card__title {
            width: auto;
            position: relative;
            top: 0;
            left: 0;
            text-align: left;
        }
    }

    .sw-card__title {
        margin-bottom: 12px;
        font-size: 18px;
        line-height: 24px;
        color: $color-darkgray-200;

        @media screen and (min-width: 1580px) {
            width: 200px;
            position: absolute;
            left: -240px;
            top: 0;
            text-align: right;
        }
    }

    .sw-card__toolbar {
        padding: 30px;
        background-color: $color-gray-100;
        border: 1px solid $color-gray-300;
    }

    &.has--header {
        .sw-card__content {
            border-top: none;
            border-radius: 0 0 $border-radius-default $border-radius-default;
        }
    }

    .sw-card__content {
        padding: 30px;
        background: $color-white;
        background-clip: padding-box;
        border-radius: $border-radius-default;
        border: 1px solid $color-gray-300;
        position: relative;

        @media screen and (max-width: 800px) {
            padding: 15px;
        }

        h1, h2, h3, h4, h5, h6 {
            font-weight: normal;
        }

        h1 {
            font-size: 24px;
        }

        h2 {
            font-size: 22px;
        }

        h3 {
            font-size: 20px;
        }

        h4, h5, h6 {
            font-size: 18px;
        }

        a.sw-card__quick-link {
            display: grid;
            grid-auto-flow: column;
            grid-column-gap: 6px;
            align-items: center;
            text-decoration: none;
            color: $color-shopware-brand-500;
            font-size: 14px;

            &:hover {
                color: $color-shopware-brand-600;
            }
        }

        .sw-tabs {
            border-width: 0 0 1px 0;
            border-bottom-right-radius: 0;
            border-bottom-left-radius: 0;
            margin-bottom: 0;

            .sw-tabs__container {
                position: relative;
                top: -5px;
            }
        }

        .sw-tabs-item {
            margin: 0;
            padding-top: 5px;
            line-height: 44px;
            border-bottom-left-radius: 0;
            border-bottom-right-radius: 0;

            &:hover {
                border-bottom-color: $color-gray-300;
            }

            &:focus {
                background-color: $color-gray-100;
                border-bottom-color: $color-gray-300;
            }

            &:active {
                border-bottom-color: $color-white;
                background-color: $color-white;
            }

            &.router-link-active {
                background-color: $color-white;
                border-bottom-color: $color-white;
            }
        }
    }
}
