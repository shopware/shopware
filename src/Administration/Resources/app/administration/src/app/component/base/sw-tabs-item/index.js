import template from './sw-tabs-item.html.twig';
import './sw-tabs-item.scss';

const { Component } = Shopware;
const types = Shopware.Utils.types;

/**
 * @private
 * @description Renders a tab item.
 * @status ready
 * @example-type static
 * @component-example
 * <sw-tabs>
 *
 *     <sw-tabs-item :route="{ name: 'sw.explore.index' }">
 *         Explore
 *     </sw-tabs-item>
 *
 *     <sw-tabs-item to="A link">
 *         My Plugins
 *     </sw-tabs-item>
 *
 * </sw-tabs>
 */
Component.register('sw-tabs-item', {
    template,

    inheritAttrs: false,

    props: {
        route: {
            type: [String, Object],
            required: false,
            default: ''
        },
        active: {
            type: Boolean,
            required: false,
            default: false
        },
        activeTab: {
            type: String,
            required: false,
            default: ''
        },
        name: {
            type: String,
            required: false,
            default: ''
        },
        hasError: {
            type: Boolean,
            required: false,
            default: false
        },
        disabled: {
            type: Boolean,
            required: false,
            default: false
        }
    },

    data() {
        return {
            isActive: false
        };
    },

    mounted() {
        this.mountedComponent();
    },

    beforeUpdate() {
        this.beforeUpdateComponent();
    },

    watch: {
        '$route'() {
            this.checkIfRouteMatchesLink();
        }
    },

    created() {
        this.createdComponent();
    },

    computed: {
        isNative() {
            return types.isEmpty(this.route);
        },

        tabsItemClasses() {
            return {
                'sw-tabs-item--active': this.isActive,
                'sw-tabs-item--has-error': this.hasError,
                'sw-tabs-item--is-disabled': this.disabled
            };
        }
    },

    methods: {
        createdComponent() {
            this.$parent.$on('new-item-active', this.checkIfActive);
            if (this.active) {
                this.isActive = true;
            }
        },

        beforeUpdateComponent() {
            this.updateActiveState();
        },

        mountedComponent() {
            this.updateActiveState();
        },
        updateActiveState() {
            this.checkIfRouteMatchesLink();
            if (this.activeTab && this.activeTab === this.name) {
                this.isActive = true;
            }
        },

        clickEvent() {
            this.$parent.setActiveItem(this);
            this.$emit('click');
        },
        checkIfActive(item) {
            this.isActive = (item.$vnode === this.$vnode);
        },
        checkIfRouteMatchesLink() {
            this.$nextTick().then(() => {
                const routeIsActive = this.$el.classList.contains('router-link-active');
                if (routeIsActive) {
                    this.$parent.setActiveItem(this);
                }
            });
        }
    }
});
