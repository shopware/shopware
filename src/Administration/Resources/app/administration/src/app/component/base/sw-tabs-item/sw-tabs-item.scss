@import "~scss/variables";

$sw-side-navigation-item-link-distance: 10px;

.sw-tabs-item {
    transition: 0.3s all ease-in-out;
    padding: $sw-side-navigation-item-link-distance ($sw-side-navigation-item-link-distance * 1.75);
    border-bottom: 2px solid $color-gray-300;
    text-decoration: none;
    white-space: nowrap;
    color: $color-darkgray-200;
    cursor: pointer;
    user-select: none;

    .sw-tabs-item__error-badge {
        position: relative;
        bottom: 2px;
        color: $color-crimson-300;
    }

    &.sw-tabs-item--active {
        color: $color-darkgray-800;

        .sw-tabs-item__error-badge {
            color: $color-crimson-500;
        }
    }

    &.sw-tabs-item--has-error {
        color: $color-crimson-300;

        &.sw-tabs-item--active {
            color: $color-crimson-500;
        }
    }

    &.sw-tabs-item--is-disabled {
        color: $color-darkgray-200;
        pointer-events: none;
    }

    &:disabled {
        color: $color-darkgray-200;
        cursor: not-allowed;
    }
}

.sw-tabs--vertical .sw-tabs-item {
    padding: $sw-side-navigation-item-link-distance;
    border-bottom: 0 solid $color-gray-300;
    border-left: 2px solid $color-gray-300;
}

.sw-tabs--vertical.sw-tabs--align-right .sw-tabs-item{
    border-left: 0;
    border-right: 2px solid $color-gray-300;
}
