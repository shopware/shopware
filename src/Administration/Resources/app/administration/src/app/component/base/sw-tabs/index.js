import template from './sw-tabs.html.twig';
import './sw-tabs.scss';

const { Component } = Shopware;
const util = Shopware.Utils;
const dom = Shopware.Utils.dom;

/**
 * @public
 * @description Renders tabs. Each item references a route or emits a custom event.
 * @status ready
 * @example-type static
 * @component-example
 * <sw-tabs>
 *
 *     <sw-tabs-item>
 *         Explore
 *     </sw-tabs-item>
 *
 *     <sw-tabs-item>
 *         My Plugins
 *     </sw-tabs-item>
 *
 * </sw-tabs>
 */
Component.register('sw-tabs', {
    template,

    props: {
        isVertical: {
            type: Boolean,
            required: false,
            default: false
        },

        small: {
            type: Boolean,
            required: false,
            default: true
        },

        alignRight: {
            type: Boolean,
            required: false,
            default: false
        },

        defaultItem: {
            type: String,
            required: false,
            default: ''
        }
    },

    data() {
        return {
            active: this.defaultItem || '',
            isScrollable: false,
            activeItem: null,
            scrollLeftPossible: false,
            scrollRightPossible: true,
            firstScroll: false,
            scrollbarOffset: ''
        };
    },

    created() {
        this.createdComponent();
    },

    mounted() {
        this.mountedComponent();
    },

    watch: {
        '$route'() {
            this.updateActiveItem();
        },

        activeTabHasErrors() {
            this.recalculateSlider();
        }
    },

    computed: {
        tabClasses() {
            return {
                'sw-tabs--vertical': this.isVertical,
                'sw-tabs--small': this.small,
                'sw-tabs--scrollable': this.isScrollable,
                'sw-tabs--align-right': this.alignRight,
                'sw-tabs--scrollbar-active': this.scrollbarOffset > 0
            };
        },

        arrowClassesLeft() {
            return {
                'sw-tabs__arrow--disabled': !this.scrollLeftPossible
            };
        },

        arrowClassesRight() {
            return {
                'sw-tabs__arrow--disabled': !this.scrollRightPossible
            };
        },

        sliderLength() {
            if (this.$children[this.activeItem]) {
                const activeChildren = this.$children[this.activeItem];
                return this.isVertical ? activeChildren.$el.offsetHeight : activeChildren.$el.offsetWidth;
            }
            return 0;
        },

        activeTabHasErrors() {
            return this.$children[this.activeItem] && this.$children[this.activeItem].hasError;
        },

        sliderClasses() {
            return { 'has--error': this.activeTabHasErrors };
        },

        sliderMovement() {
            if (this.$children[this.activeItem]) {
                const activeChildren = this.$children[this.activeItem];
                return this.isVertical ? activeChildren.$el.offsetTop : activeChildren.$el.offsetLeft;
            }
            return 0;
        },

        sliderStyle() {
            if (this.isVertical) {
                return `
                    transform: translate(0, ${this.sliderMovement}px) rotate(${this.alignRight ? '-90deg' : '90deg'});
                    width: ${this.sliderLength}px;
                `;
            }

            return `
                transform: translate(${this.sliderMovement}px, 0) rotate(0deg);
                width: ${this.sliderLength}px;
                bottom: ${this.scrollbarOffset}px;
            `;
        },

        tabContentStyle() {
            return {
                'padding-bottom': `${this.scrollbarOffset}px`
            };
        }
    },

    methods: {
        mountedComponent() {
            const tabContent = this.$refs.swTabContent;

            tabContent.addEventListener('scroll', util.throttle(() => {
                const rightEnd = tabContent.scrollWidth - tabContent.offsetWidth;
                const leftDistance = tabContent.scrollLeft;

                this.scrollRightPossible = !(rightEnd - leftDistance < 5);
                this.scrollLeftPossible = !(leftDistance < 5);
            }, 100));

            this.checkIfNeedScroll();
            this.addScrollbarOffset();

            const that = this;
            this.$device.onResize({
                listener() {
                    that.checkIfNeedScroll();
                    that.addScrollbarOffset();
                },
                component: this
            });
            this.recalculateSlider();
        },

        recalculateSlider() {
            window.setTimeout(() => {
                const activeItem = this.activeItem;
                this.activeItem = null;
                this.activeItem = activeItem;
            }, 0);
        },

        createdComponent() {
            this.updateActiveItem();
        },

        updateActiveItem() {
            this.$nextTick().then(() => {
                const firstActiveTabItem = this.$children.find((child) => {
                    return child.$el.nodeType === 1 && child.$el.classList.contains('sw-tabs-item--active');
                });

                if (!firstActiveTabItem) {
                    return;
                }

                this.activeItem = this.$children.indexOf(firstActiveTabItem);
                if (!this.firstScroll) {
                    this.scrollToItem(firstActiveTabItem);
                }
                this.firstScroll = true;
            });
        },

        scrollTo(direction) {
            if (!['left', 'right'].includes(direction)) {
                return;
            }

            const tabContent = this.$refs.swTabContent;
            const tabContentWidth = tabContent.offsetWidth;

            if (direction === 'right') {
                tabContent.scrollLeft += (tabContentWidth / 2);
                return;
            }
            tabContent.scrollLeft += -(tabContentWidth / 2);
        },

        checkIfNeedScroll() {
            const tabContent = this.$refs.swTabContent;
            this.isScrollable = tabContent.scrollWidth !== tabContent.offsetWidth;
        },

        setActiveItem(item) {
            this.$emit('new-item-active', item);
            this.active = item.name;
            this.updateActiveItem();
        },

        scrollToItem(item) {
            const tabContent = this.$refs.swTabContent;
            const tabContentWidth = tabContent.offsetWidth;
            const itemOffset = item.$el.offsetLeft;
            const itemWidth = item.$el.clientWidth;

            if ((tabContentWidth / 2) < itemOffset) {
                const scrollWidth = itemOffset - (tabContentWidth / 2) + (itemWidth / 2);
                tabContent.scrollLeft = scrollWidth;
            }
        },

        addScrollbarOffset() {
            this.scrollbarOffset = dom.getScrollbarHeight(this.$refs.swTabContent);
        }
    }
});
