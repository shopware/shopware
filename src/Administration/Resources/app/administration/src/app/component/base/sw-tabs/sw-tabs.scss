@import "~scss/variables";

.sw-tabs {
    position: relative;

    .sw-tabs__content {
        position: relative;
        display: flex;
        overflow: auto;
        scroll-behavior: smooth;

        &::-webkit-scrollbar {
            display: none;
        }

        &::before {
            content: "";
            display: block;
            position: absolute;
            bottom: 0;
            width: 100%;
            height: 2px;
            background-color: $color-gray-300;
        }
    }

    .sw-tabs__slider {
         transform-origin: top left;
         transition: 0.2s all ease-in-out;
         position: absolute;
         bottom: 0;
         height: 2px;
         background-color: $color-shopware-brand-500;

        &.has--error {
            background-color: $color-crimson-500;
        }
    }

    .sw-tabs__arrow {
        transition: 0.3s all ease-in-out;
        position: absolute;
        left: 0;
        top: 21px;
        transform: translate(0, -50%);
        border: none;
        background-color: transparent;

        &.sw-tabs__arrow--right {
            left: auto;
            right: 0;
        }

        &.sw-tabs__arrow--disabled {
            cursor: default !important;
            opacity: 0.25;
        }

        .sw-icon {
            transition: 0.3s all ease;
        }

        &:focus {
            outline: none;
        }

        &:hover {
            outline: none;
            cursor: pointer;

            .sw-icon {
                color: $color-shopware-brand-500;
            }
        }
    }

    .sw-tabs__custom-content {
        padding: 20px 0;
    }


    &.sw-tabs--vertical {
        .sw-tabs__content {
            flex-direction: column;

            .sw-tabs__slider {
                bottom: auto;
                top: 0;
                left: 2px;
            }

            &::before {
                display: none;
            }
        }

        &.sw-tabs--align-right {
            .sw-tabs__content {
                .sw-tabs-item {
                    text-align: right;
                }

                .sw-tabs__slider {
                    transform-origin: bottom right;
                    left: auto;
                    right: 0;
                    top: -2px;
                }
            }
        }
    }

    &.sw-tabs--small {
        max-width: 800px;
        margin: 0 auto 15px auto;
    }

    &.sw-tabs--scrollable {
        padding: 0 20px;
    }

    &.sw-tabs--scrollbar-active {
        .sw-tabs__content {
            &::before {
                display: none;
            }
        }
    }
}
