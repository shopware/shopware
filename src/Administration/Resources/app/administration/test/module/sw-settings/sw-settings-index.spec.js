import { shallowMount } from '@vue/test-utils';
import 'src/module/sw-settings/page/sw-settings-index';
import 'src/module/sw-settings/component/sw-settings-item';
import 'src/app/component/base/sw-card';
import 'src/app/component/base/sw-tabs';
import 'src/app/component/base/sw-tabs-item';


function createWrapper(privileges = [
    'store.viewer',
    'user.viewer',
    'foo.viewer',
    'snippet.viewer',
    'store.viewer',
    'listing.viewer',
    'shipping.viewer'
]) {
    const settingsItemsMock = [
        {
            group: 'system',
            to: 'sw.settings.store.index',
            icon: 'default-device-laptop',
            id: 'sw-settings-store',
            name: 'settings-store',
            label: 'c',
            privilege: 'store.viewer'
        },
        {
            group: 'system',
            to: 'sw.settings.user.list',
            icon: 'default-avatar-single',
            id: 'sw-settings-user',
            name: 'settings-user',
            label: 'a',
            privilege: 'user.viewer'
        },
        {
            group: 'system',
            to: 'sw.settings.foo.list',
            icon: 'default-avatar-single',
            id: 'sw-settings-foo',
            name: 'settings-foo',
            label: 'b',
            privilege: 'foo.viewer'
        },
        {
            group: 'shop',
            to: 'sw.settings.snippet.index',
            icon: 'default-object-globe',
            id: 'sw-settings-snippet',
            name: 'settings-snippet',
            label: 'h',
            privilege: 'snippet.viewer'
        },
        {
            group: 'shop',
            to: 'sw.settings.listing.index',
            icon: 'default-symbol-products',
            id: 'sw-settings-listing',
            name: 'settings-listing',
            label: 's',
            privilege: 'listing.viewer'
        },
        {
            group: 'shop',
            to: 'sw.settings.shipping.index',
            icon: 'default-package-open',
            id: 'sw-settings-shipping',
            name: 'settings-shipping',
            label: 'a',
            privilege: 'shipping.viewer'
        }
    ];

    settingsItemsMock.forEach((settingsItem) => {
        Shopware.State.commit('settingsItems/addItem', settingsItem);
    });

    return shallowMount(Shopware.Component.build('sw-settings-index'), {
        stubs: {
            'sw-page': {
                template: '<div><slot name="content"></slot></div>'
            },
            'sw-card-view': {
                template: '<div class="sw-card-view"><slot></slot></div>'
            },
            'sw-tabs': Shopware.Component.build('sw-tabs'),
            'sw-tabs-item': Shopware.Component.build('sw-tabs-item'),
            'sw-card': {
                template: '<div class="sw-card"><slot></slot></div>'
            },
            'sw-settings-item': Shopware.Component.build('sw-settings-item'),
            'router-link': {
                template: '<a><slot></slot></a>'
            },
            'sw-icon': {
                template: '<span></span>'
            }
        },
        provide: {
            acl: {
                can: (key) => {
                    if (!key) return true;

                    return privileges.includes(key);
                }
            }
        },
        mocks: {
            $tc: (value) => value,
            $device: { onResize: () => {} }
        }
    });
}

describe('module/sw-settings/page/sw-settings-index', () => {
    beforeEach(() => {
        Shopware.State.get('settingsItems').settingsGroups = {};
    });

    it('should be a Vue.js component', async () => {
        const wrapper = createWrapper();
        expect(wrapper.vm).toBeTruthy();
    });

    it('should contain any settings items', async () => {
        const wrapper = createWrapper();
        expect(wrapper.vm.settingsGroups).not.toEqual({});
    });

    it('should return settings items alphabetically sorted', async () => {
        const wrapper = createWrapper();
        const settingsGroups = Object.entries(wrapper.vm.settingsGroups);

        settingsGroups.forEach(([, settingsItems]) => {
            settingsItems.forEach((settingsItem, index) => {
                let elementsSorted = true;

                if (index < settingsItems.length - 1) {
                    elementsSorted = (
                        settingsItems[index].label.localeCompare(settingsItems[index + 1].label) === -1
                    );
                }

                expect(elementsSorted).toEqual(true);
            });
        });
    });

    it('should render correctly', async () => {
        const wrapper = createWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should render settings items in alphabetical order', async () => {
        const wrapper = createWrapper();
        const settingsGroups = Object.entries(wrapper.vm.settingsGroups);

        settingsGroups.forEach(([settingsGroup, settingsItems]) => {
            const settingsGroupWrapper = wrapper.find(`#sw-settings__content-grid-${settingsGroup}`);
            const settingsItemsWrappers = settingsGroupWrapper.findAll('.sw-settings-item');

            // check, that all settings items were rendered
            expect(settingsItemsWrappers.length).toEqual(settingsItems.length);

            // check, that settings items were rendered in alphabetical order
            settingsItemsWrappers.wrappers.forEach((settingsItemsWrapper, index) => {
                expect(settingsItemsWrapper.attributes().id).toEqual(settingsItems[index].id);
            });
        });
    });

    it('should render settings items in alphabetical order with updated items', async () => {
        const settingsItemToAdd = {
            group: 'shop',
            to: 'sw.bar.index',
            icon: 'bar',
            id: 'sw-settings-bar',
            name: 'settings-bar',
            label: 'b'
        };

        Shopware.State.commit('settingsItems/addItem', settingsItemToAdd);

        const wrapper = createWrapper();
        const settingsGroups = Object.entries(wrapper.vm.settingsGroups);

        settingsGroups.forEach(([settingsGroup, settingsItems]) => {
            const settingsGroupWrapper = wrapper.find(`#sw-settings__content-grid-${settingsGroup}`);
            const settingsItemsWrappers = settingsGroupWrapper.findAll('.sw-settings-item');

            expect(settingsItemsWrappers.length).toEqual(settingsItems.length);

            settingsItemsWrappers.wrappers.forEach((settingsItemsWrapper, index) => {
                expect(settingsItemsWrapper.attributes().id).toEqual(settingsItems[index].id);
            });
        });
    });

    it('should add the setting to the settingsGroups in store', async () => {
        const settingsItemToAdd = {
            group: 'shop',
            to: 'sw.bar.index',
            icon: 'bar',
            id: 'sw-settings-bar',
            name: 'settings-bar',
            label: 'b'
        };

        Shopware.State.commit('settingsItems/addItem', settingsItemToAdd);

        const wrapper = createWrapper();

        const settingsGroups = wrapper.vm.settingsGroups.shop;
        const barSetting = settingsGroups.find(setting => setting.id === 'sw-settings-bar');

        expect(barSetting).toBeDefined();
    });

    it('should show the setting with the privileges', async () => {
        const settingsItemToAdd = {
            privilege: 'system.foo_bar',
            group: 'shop',
            to: 'sw.bar.index',
            icon: 'bar',
            id: 'sw-settings-bar',
            name: 'settings-bar',
            label: 'b'
        };

        Shopware.State.commit('settingsItems/addItem', settingsItemToAdd);

        const wrapper = createWrapper('system.foo_bar');

        const settingsGroups = wrapper.vm.defaultSettingsGroups.shop;
        const barSetting = settingsGroups.find(setting => setting.id === 'sw-settings-bar');

        expect(barSetting).toBeDefined();
    });

    it('should not show the setting with the privileges', async () => {
        const settingsItemToAdd = {
            privilege: 'system.foo_bar',
            group: 'shop',
            to: 'sw.bar.index',
            icon: 'bar',
            id: 'sw-settings-bar',
            name: 'settings-bar',
            label: 'b'
        };

        Shopware.State.commit('settingsItems/addItem', settingsItemToAdd);

        const wrapper = createWrapper();

        const settingsGroups = wrapper.vm.defaultSettingsGroups.shop;
        const barSetting = settingsGroups.find(setting => setting.id === 'sw-settings-bar');

        expect(barSetting).toBeUndefined();
    });

    it('should hide icon background when backgroundEnabled is false', async () => {
        const settingsItemToAdd = {
            group: 'shop',
            to: 'sw.bar.index',
            icon: 'bar',
            id: 'settings-background-disabled',
            name: 'settings-background-disabled',
            label: 'b',
            backgroundEnabled: false
        };

        Shopware.State.commit('settingsItems/addItem', settingsItemToAdd);

        const wrapper = createWrapper();

        const settingsGroups = Object.entries(wrapper.vm.settingsGroups);

        settingsGroups.forEach(([settingsGroup, settingsItems]) => {
            const settingsGroupWrapper = wrapper.find(`#sw-settings__content-grid-${settingsGroup}`);
            const settingsItemsWrappers = settingsGroupWrapper.findAll('.sw-settings-item');

            settingsItemsWrappers.wrappers.forEach((settingsItemsWrapper, index) => {
                const iconClasses = settingsItemsWrapper.find('.sw-settings-item__icon').attributes().class;

                if (settingsItems[index].backgroundEnabled === false) {
                    expect(iconClasses).not.toContain('background--enabled');
                } else {
                    expect(iconClasses).toContain('background--enabled');
                }
            });
        });
    });

    it('should hide the tab when user has no access to any settings inside the tab', async () => {
        const settingsItemToAdd = {
            privilege: 'system.foo_bar',
            group: 'shop',
            to: 'sw.bar.index',
            icon: 'bar',
            id: 'sw-settings-bar',
            name: 'settings-bar',
            label: 'b'
        };

        Shopware.State.commit('settingsItems/addItem', settingsItemToAdd);

        const wrapper = await createWrapper('system.foo_bar');

        const systemTab = wrapper.find('.sw-settings__tab-system');
        const shopTab = wrapper.find('.sw-settings__tab-shop');

        expect(systemTab.exists()).toBe(false);
        expect(shopTab.exists()).toBe(true);
    });

    it('should hide the tab when user has no access to any settings inside the tab', async () => {
        const settingsItemToAdd = {
            privilege: 'system.foo_bar',
            group: 'system',
            to: 'sw.bar.index',
            icon: 'bar',
            id: 'sw-settings-bar',
            name: 'settings-bar',
            label: 'b'
        };

        Shopware.State.commit('settingsItems/addItem', settingsItemToAdd);

        const wrapper = await createWrapper('system.foo_bar');

        const systemTab = wrapper.find('.sw-settings__tab-system');
        const shopTab = wrapper.find('.sw-settings__tab-shop');

        expect(systemTab.exists()).toBe(true);
        expect(shopTab.exists()).toBe(false);
    });
});
