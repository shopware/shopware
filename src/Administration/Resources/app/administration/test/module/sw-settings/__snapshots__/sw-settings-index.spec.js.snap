// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module/sw-settings/page/sw-settings-index should render correctly 1`] = `
<div
  class="sw-settings-index"
>
  <div
    class="sw-card-view"
  >
    <div
      class="sw-tabs sw-settings__tabs sw-tabs--vertical"
    >
      <div
        class="sw-tabs__content"
      >
        <a
          class="sw-tabs-item sw-settings__tab-shop"
          title="sw-settings.index.tabShop"
        >
          
                                        sw-settings.index.tabShop
                                    
        </a>
         
        <a
          class="sw-tabs-item sw-settings__tab-system"
          title="sw-settings.index.tabSystem"
        >
          
                                        sw-settings.index.tabSystem
                                    
        </a>
         
        <!---->
         
        <span
          class="sw-tabs__slider"
          style="transform: translate(0, 0px) rotate(90deg); width: 0px;"
        />
      </div>
       
      <!---->
       
      <!---->
       
      <div
        class="sw-tabs__custom-content"
      >
        <h1
          class="sw-settings__content-header"
        >
          sw-settings.index.title
        </h1>
         
        <div
          class="sw-card sw-settings__card"
        >
          <div
            class="sw-settings__content-grid"
            id="sw-settings__content-grid-shop"
          >
            <a
              class="sw-settings-item"
              id="sw-settings-shipping"
              to="[object Object]"
            >
              <div
                class="sw-settings-item__icon background--enabled"
              >
                <span
                  name="default-package-open"
                />
              </div>
               
              <div
                class="sw-settings-item__label"
              >
                a
              </div>
            </a>
            <a
              class="sw-settings-item"
              id="sw-settings-snippet"
              to="[object Object]"
            >
              <div
                class="sw-settings-item__icon background--enabled"
              >
                <span
                  name="default-object-globe"
                />
              </div>
               
              <div
                class="sw-settings-item__label"
              >
                h
              </div>
            </a>
            <a
              class="sw-settings-item"
              id="sw-settings-listing"
              to="[object Object]"
            >
              <div
                class="sw-settings-item__icon background--enabled"
              >
                <span
                  name="default-symbol-products"
                />
              </div>
               
              <div
                class="sw-settings-item__label"
              >
                s
              </div>
            </a>
             
            <div />
             
            <!---->
          </div>
          <div
            class="sw-settings__content-grid"
            id="sw-settings__content-grid-system"
            style="display: none;"
          >
            <a
              class="sw-settings-item"
              id="sw-settings-user"
              to="[object Object]"
            >
              <div
                class="sw-settings-item__icon background--enabled"
              >
                <span
                  name="default-avatar-single"
                />
              </div>
               
              <div
                class="sw-settings-item__label"
              >
                a
              </div>
            </a>
            <a
              class="sw-settings-item"
              id="sw-settings-foo"
              to="[object Object]"
            >
              <div
                class="sw-settings-item__icon background--enabled"
              >
                <span
                  name="default-avatar-single"
                />
              </div>
               
              <div
                class="sw-settings-item__label"
              >
                b
              </div>
            </a>
            <a
              class="sw-settings-item"
              id="sw-settings-store"
              to="[object Object]"
            >
              <div
                class="sw-settings-item__icon background--enabled"
              >
                <span
                  name="default-device-laptop"
                />
              </div>
               
              <div
                class="sw-settings-item__label"
              >
                c
              </div>
            </a>
             
            <!---->
             
            <div />
          </div>
           
          <div
            class="sw-settings__content-grid"
            style="display: none;"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
